### PR TITLE
Use model.from_pretrained for DataParallel also

### PR DESCRIPTION
**Pull request by shaie**
When training on multiple GPUs, the code wraps a model with torch.nn.DataParallel. However if the model has custom from_pretrained logic, it does not get applied during load_best_model_at_end.

This commit uses the underlying model during load_best_model_at_end, and re-wraps the loaded model with DataParallel.

If you choose to reject this change, then could you please move the this logic to a function, e.g. def load_best_model_checkpoint(best_model_checkpoint) or something, so that it can be overridden?

# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -808,8 +808,15 @@ def train(self, model_path: Optional[str] = None, trial: Union["optuna.Trial", D
...
             )
+            is_data_parallel = False
+            if isinstance(model, torch.nn.DataParallel):
+                model = model.module
+                is_data_parallel = True
             if isinstance(model, PreTrainedModel):
                 self.model = model.from_pretrained(self.state.best_model_checkpoint)
+                if is_data_parallel:
+                    # re-wrap with DataParallel
+                    self.model = torch.nn.DataParallel(self.model)
```
Comment (sgugger) [inline src/transformers/trainer.py:819]: The model attribute of the `Learner` is never wrapped in `DataParallel` or the like, it stays a reference to the original model so these lines are not necessary.
With the same reasoning, I think the only change to make is on the line above this comment (L816) and replace
```
self.model = model.from_pretrained(self.state.best_model_checkpoint)
```
by
```
self.model = self.model.from_pretrained(self.state.best_model_checkpoint)
```

**Review by sgugger — COMMENTED** [review]
I understand the problem but I disagree with the solution you picked: the `model` attribute of the `Learner` is always a reference to the original (unwrapped) model. I suggested a change to use that for the fix, would you mind applying it?
Thanks!

**Review by sgugger — APPROVED** [review]
Perfect, thanks a lot!

**Comment by sgugger** [comment]
Oh, looks like there is a last code-style issue to fix. Could you run `make style` on your branch? Then we can merge this.

**Comment by shaie** [comment]
I don't have `make` installed 😄 , what is the style issue? Wonder what style issue can go wrong in such a simple patch. The only thing we added is `self.` in those 2 lines

**Comment by shaie** [comment]
`check_code_quality` complains about `finetune.py`, but it's not modified by this patch

**Comment by sgugger** [comment]
Weird indeed. Will merge and fix if the issue persists.

**Pull request merged**